src/main/java/com/google/api/codegen/util/Name.java

```diff
@@ -0,0 +1,204 @@
+/* Copyright 2016 Google Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.api.codegen.util;
+
+import com.google.api.client.util.Joiner;
+import com.google.common.base.CaseFormat;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Name represents an identifier name which is casing-aware.
+ */
+public class Name {
+  private List<NamePiece> namePieces;
+
+  /**
+   * Creates a Name from a sequence of lower-underscore strings.
+   *
+   * @throws IllegalArgumentException if any of the strings contain any characters that are not
+   * lower case or underscores.
+   */
+  public static Name from(String... pieces) {
```

Comment:
Why not name this lowerUnderscore, similar to lowerCamel below, and dotted in NamePath object